src/qibolab/_core/instruments/keysight/qcs.py

```diff
@@ -0,0 +1,172 @@
+"""Qibolab driver for Keysight QCS instrument set."""
+
+from collections import defaultdict
+from functools import reduce
+from typing import ClassVar, Optional
+
+import numpy as np
+from keysight import qcs
+
+from qibolab._core.components import AcquisitionChannel, Config, DcChannel, IqChannel
+from qibolab._core.execution_parameters import AveragingMode, ExecutionParameters
+from qibolab._core.identifier import ChannelId, Result
+from qibolab._core.instruments.abstract import Controller
+from qibolab._core.pulses import PulseId
+from qibolab._core.sequence import InputOps, PulseSequence
+from qibolab._core.sweeper import ParallelSweepers
+from qibolab._core.unrolling import Bounds
+
+from .pulse import (
+    process_acquisition_channel_pulses,
+    process_dc_channel_pulses,
+    process_iq_channel_pulses,
+)
+from .results import fetch_result, parse_result
+from .sweep import process_sweepers
+
+NS_TO_S = 1e-9
+
+BOUNDS = Bounds(
+    waveforms=1,
+    readout=1,
+    instructions=1,
+)
+
+__all__ = ["KeysightQCS"]
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
+    # Map of Qibolab channel IDs to QCS virtual channels
+    virtual_channel_map: dict[ChannelId, qcs.Channels]
+    # Map of QCS virtual acquisition channels to QCS state classifiers
+    classifier_map: Optional[dict[qcs.Channels, qcs.MinimumDistanceClassifier]] = {}
```

Comment:
Again looking at https://github.com/qiboteam/qibolab_platforms_nqch/blob/qibolab-0.2/iqm5q/platform.py, the main issue I see here is that you are storing calibrated parameters of the classifier in platform.py. Normally, we store all calibrated parameters in parameters.json because this is much easier to automatically update after running calibration routines and we already have the related infrastructure for that in qibocal.

Ideally, we should store these parameters in the `AcquisitionConfig`, which is loaded from the JSON, like we do for the `threshold` and `iq_angle` in other instruments. If our default `AcquisitionConfig` is not sufficient or convenient for your classifier, you can define an instrument specific config similar to https://github.com/qiboteam/qibolab/blob/3a48f8338d1933cee58777029cb9b750f6e1739a/src/qibolab/_core/instruments/qm/components/configs.py#L48
save your parameters with the preferred names there and then instantiate the `qcs.MinimumDistanceClassifier` inside the driver using these parameters. This way they will be stored in the JSON and qibolab will automatically handle their loading and saving.

Reply:
Implemented in https://github.com/qiboteam/qibolab/pull/944/commits/3a29ffaaa91380e530b8862c85b271be89724d2b